InnerEye/ML/run_ml.py

```diff
@@ -251,75 +270,65 @@ def run(self) -> None:
 
         # We specify the ModelProcessing as DEFAULT here even if the run_recovery points to an ensemble run, because
         # the current run is a single one. See the documentation of ModelProcessing for more details.
-        best_epoch = self.run_inference_and_register_model(checkpoint_handler, ModelProcessing.DEFAULT)
+        self.run_inference_and_register_model(checkpoint_handler, ModelProcessing.DEFAULT)
 
-        # Generate report
-        if best_epoch:
-            Runner.generate_report(self.model_config, best_epoch, ModelProcessing.DEFAULT)
-        elif self.model_config.is_scalar_model and len(self.model_config.get_test_epochs()) == 1:
-            # We don't register scalar models but still want to create a report if we have run inference.
-            Runner.generate_report(self.model_config, self.model_config.get_test_epochs()[0], ModelProcessing.DEFAULT)
+        if self.model_config.generate_report:
+            self.generate_report(ModelProcessing.DEFAULT)
+
+        # If this is an cross validation run, and the present run is child run 0, then wait for the sibling runs,
+        # build the ensemble model, and write a report for that.
+        if self.model_config.number_of_cross_validation_splits > 0:
+            if self.model_config.should_wait_for_other_cross_val_child_runs():
+                self.wait_for_runs_to_finish()
+                self.create_ensemble_model()
 
     def run_inference_and_register_model(self, checkpoint_handler: CheckpointHandler,
-                                         model_proc: ModelProcessing) -> Optional[int]:
+                                         model_proc: ModelProcessing) -> None:
         """
         Run inference as required, and register the model, but not necessarily in that order:
         if we can identify the epoch to register at without running inference, we register first.
         :param checkpoint_handler: Checkpoint handler object to find checkpoint paths for model initialization
         :param model_proc: whether we are running an ensemble model from within a child run with index 0. If we are,
         then outputs will be written to OTHER_RUNS/ENSEMBLE under the main outputs directory.
         """
-        registration_epoch = self.decide_registration_epoch_without_evaluating()
-        if registration_epoch is not None:
-            model_description = f"Registering model for epoch {registration_epoch} without considering metrics."
-            checkpoint_paths = checkpoint_handler.get_checkpoint_paths_from_epoch_or_fail(registration_epoch)
+
+        if self.should_register_model():
```

Comment:
Should the function should_register_model return False for offline runs? We can remove multiple checks in register_model_for_epoch and register_segmentation_model.

Reply:
good idea, done.